File: app/src/main/java/com/example/voicenavigation/voice/VoiceCommandInterpreter.java
package com.example.voicenavigation.voice;

import android.util.Log;

import androidx.annotation.NonNull;

import java.util.Arrays;
import java.util.List;

/**
 * 语音指令意图解析器。
 *
 * <p>将语音识别得到的文本解析为结构化的 {@link VoiceCommand}。
 * 采用<strong>本地关键词匹配</strong>策略，无需联网即可实时响应，
 * 适合盲人用户对响应速度的敏感需求。</p>
 *
 * <p>架构上预留了接入云端 LLM Function Calling 的扩展点：
 * 可将 {@link #interpret(String)} 替换为远程 LLM API 调用。</p>
 */
public class VoiceCommandInterpreter {

    private static final String TAG = "VoiceCommandInterpreter";

    // ==================== 意图关键词库 ====================

    /** 停止导航 */
    private static final List<String> STOP_NAV_KEYWORDS = Arrays.asList(
            "停止导航", "结束导航", "关闭导航", "取消导航", "不要导航了",
            "不导航了", "退出导航", "停下导航"
    );

    /** 开始避障 */
    private static final List<String> START_OBSTACLE_KEYWORDS = Arrays.asList(
            "开始避障", "打开避障", "启动避障", "避障模式", "检测障碍物",
            "障碍物检测", "进入避障", "开启避障", "避障"
    );

    /** 停止避障 */
    private static final List<String> STOP_OBSTACLE_KEYWORDS = Arrays.asList(
            "停止避障", "结束避障", "关闭避障", "退出避障", "取消避障",
            "不要避障了", "不避障了", "停下避障"
    );

    /** 我在哪里 */
    private static final List<String> WHERE_AM_I_KEYWORDS = Arrays.asList(
            "我在哪里", "我在哪儿", "当前位置", "我的位置", "这是哪里",
            "这是哪儿", "我在什么地方", "现在在哪"
    );

    /** 重复播报 */
    private static final List<String> REPEAT_KEYWORDS = Arrays.asList(
            "重复", "再说一遍", "重新播报", "再讲一遍", "再听一遍",
            "重复一遍", "刚才说什么", "再说一次"
    );

    /** 预览路线 */
    private static final List<String> PREVIEW_KEYWORDS = Arrays.asList(
            "预览路线", "路线预览", "行前预览", "预览导航", "看路线",
            "查看路线"
    );

    /** 查询状态 */
    private static final List<String> STATUS_KEYWORDS = Arrays.asList(
            "查询状态", "现在什么情况", "当前状态", "状态怎么样",
            "怎么样了", "什么情况", "在干什么"
    );

    /** 导航前缀：文本以这些词开头时，视为导航意图 */
    private static final List<String> NAVIGATE_PREFIXES = Arrays.asList(
            "导航到", "导航去", "带我去", "我要去", "我想去",
            "请带我去", "带我去一下", "去一下"
    );

    /** 导航后缀：文本以这些词结尾时，视为导航意图 */
    private static final List<String> NAVIGATE_SUFFIXES = Arrays.asList(
            "怎么走", "怎么去", "在哪里", "在哪儿", "的位置",
            "怎么去啊", "怎么走啊"
    );

    /**
     * 解析语音文本，返回结构化的语音指令。
     *
     * @param text 语音识别得到的原始文本
     * @return 解析后的 VoiceCommand，不会返回 null
     */
    @NonNull
    public VoiceCommand interpret(@NonNull String text) {
        String cleaned = cleanText(text);
        Log.d(TAG, "Interpreting: \"" + cleaned + "\"");

        // 1. 精确匹配：停止导航（优先级高，避免与"导航到"混淆）
        if (containsAny(cleaned, STOP_NAV_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.STOP_NAVIGATION, null, cleaned);
        }

        // 2. 精确匹配：停止避障
        if (containsAny(cleaned, STOP_OBSTACLE_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.STOP_OBSTACLE_AVOIDANCE, null, cleaned);
        }

        // 3. 精确匹配：开始避障
        if (containsAny(cleaned, START_OBSTACLE_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.START_OBSTACLE_AVOIDANCE, null, cleaned);
        }

        // 4. 精确匹配：我在哪里
        if (containsAny(cleaned, WHERE_AM_I_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.WHERE_AM_I, null, cleaned);
        }

        // 5. 精确匹配：重复播报
        if (containsAny(cleaned, REPEAT_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.REPEAT_LAST, null, cleaned);
        }

        // 6. 精确匹配：预览路线
        if (containsAny(cleaned, PREVIEW_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.PREVIEW_ROUTE, null, cleaned);
        }

        // 7. 精确匹配：查询状态
        if (containsAny(cleaned, STATUS_KEYWORDS)) {
            return new VoiceCommand(VoiceCommand.Type.QUERY_STATUS, null, cleaned);
        }

        // 8. 导航意图：提取目的地
        String destination = extractDestination(cleaned);
        if (destination != null && !destination.isEmpty()) {
            Log.d(TAG, "Parsed NAVIGATE_TO destination: " + destination);
            return new VoiceCommand(VoiceCommand.Type.NAVIGATE_TO, destination, cleaned);
        }

        // 9.  fallback：作为普通文本搜索
        if (cleaned.length() >= 2) {
            Log.d(TAG, "Falling back to TEXT_SEARCH: " + cleaned);
            return new VoiceCommand(VoiceCommand.Type.TEXT_SEARCH, cleaned, cleaned);
        }

        Log.d(TAG, "Unknown command: " + cleaned);
        return new VoiceCommand(VoiceCommand.Type.UNKNOWN, null, cleaned);
    }

    /**
     * 清洗语音文本：去首尾标点、去语气词。
     */
    @NonNull
    private String cleanText(@NonNull String text) {
        String cleaned = text.trim()
                .replaceAll("^[。，、！；：,.!?;:]+", "")
                .replaceAll("[。，、！；：,.!?;:]+$", "")
                .replaceAll("^[吧吗呢啊]+", "")
                .replaceAll("[吧吗呢啊]+$", "");
        return cleaned.trim();
    }

    /**
     * 判断文本是否包含关键词列表中的任意一个。
     */
    private boolean containsAny(@NonNull String text, @NonNull List<String> keywords) {
        for (String keyword : keywords) {
            if (text.contains(keyword)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 从文本中提取导航目的地。
     * 支持前缀形式（"导航到天安门"）和后缀形式（"天安门怎么走"）。
     *
     * @return 提取到的目的地，如果不是导航意图则返回 null
     */
    private String extractDestination(@NonNull String text) {
        // 前缀匹配："导航到XX"、"带我去XX" 等
        for (String prefix : NAVIGATE_PREFIXES) {
            if (text.startsWith(prefix)) {
                String dest = text.substring(prefix.length()).trim()
                        .replaceAll("^[的吧吗呢啊]+", "")
                        .replaceAll("[的吧吗呢啊]+$", "")
                        .trim();
                if (!dest.isEmpty()) {
                    return dest;
                }
            }
        }

        // 后缀匹配："XX怎么走"、"XX怎么去" 等
        for (String suffix : NAVIGATE_SUFFIXES) {
            if (text.endsWith(suffix)) {
                String dest = text.substring(0, text.length() - suffix.length()).trim()
                        .replaceAll("^[的吧吗呢啊]+", "")
                        .replaceAll("[的吧吗呢啊]+$", "")
                        .trim();
                if (!dest.isEmpty()) {
                    return dest;
                }
            }
        }

        // 特殊处理：单独的 "去XX"
        if (text.startsWith("去") && text.length() > 1) {
            String dest = text.substring(1).trim()
                    .replaceAll("^[的吧吗呢啊]+", "")
                    .replaceAll("[的吧吗呢啊]+$", "")
                    .trim();
            if (!dest.isEmpty() && dest.length() >= 2) {
                return dest;
            }
        }

        // 特殊处理：单独的 "到XX"
        if (text.startsWith("到") && text.length() > 1) {
            String dest = text.substring(1).trim()
                    .replaceAll("^[的吧吗呢啊]+", "")
                    .replaceAll("[的吧吗呢啊]+$", "")
                    .trim();
            if (!dest.isEmpty() && dest.length() >= 2) {
                return dest;
            }
        }

        return null;
    }
}

File: app/src/main/java/com/example/voicenavigation/voice/VoiceCommandInterpreter.kt
package com.example.voicenavigation.voice

import android.content.Context
import android.util.Log
import org.json.JSONObject

/**
 * 语音指令意图解析器。
 *
 * 将语音识别得到的文本解析为 [VoiceCommand]。
 * 采用**本地关键词匹配**策略，无需联网即可实时响应。
 *
 * 关键词从 `assets/voice_keywords.json` 加载，新增/修改关键词只需编辑 JSON，无需重编译。
 */
class VoiceCommandInterpreter(private val context: Context? = null) {

    companion object {
        private const val TAG = "VoiceCommandInterpreter"
        private const val KEYWORDS_FILE = "voice_keywords.json"
    }

    private val keywords: Map<String, List<String>> = loadKeywords()

    private fun loadKeywords(): Map<String, List<String>> {
        val ctx = context ?: return emptyMap()
        return try {
            val json = ctx.assets.open(KEYWORDS_FILE).bufferedReader().readText()
            val obj = JSONObject(json)
            obj.keys().asSequence().associateWith { key ->
                val arr = obj.getJSONArray(key)
                (0 until arr.length()).map { arr.getString(it) }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load $KEYWORDS_FILE, using empty keywords", e)
            emptyMap()
        }
    }

    private fun getKeywords(key: String): List<String> = keywords[key] ?: emptyList()

    /**
     * 解析语音文本，返回结构化的语音指令。
     */
    fun interpret(text: String): VoiceCommand {
        val cleaned = cleanText(text)
        Log.d(TAG, "Interpreting: \"$cleaned\"")

        // 0. 金造村路线选择（最高优先级：精确 + 模糊）
        if (fuzzyMatch(cleaned, getKeywords("jinzao_route_qiyi"))) {
            Log.d(TAG, "jinzao_route_qiyi matched (fuzzy): \"$cleaned\"")
            return VoiceCommand(VoiceCommand.Type.JINZAO_ROUTE_QIYI, "起义广场", cleaned)
        }
        if (fuzzyMatch(cleaned, getKeywords("jinzao_route_silingbu"))) {
            Log.d(TAG, "jinzao_route_silingbu matched (fuzzy): \"$cleaned\"")
            return VoiceCommand(VoiceCommand.Type.JINZAO_ROUTE_SILINGBU, "司令部旧址", cleaned)
        }

        // 1. 金造村参观游览（触发引导入口）
        if (containsAny(cleaned, getKeywords("jinzao_tour"))) {
            return VoiceCommand(VoiceCommand.Type.JINZAO_TOUR, "金造村", cleaned)
        }

        // 3. 停止导航
        if (containsAny(cleaned, getKeywords("stop_navigation"))) {
            return VoiceCommand(VoiceCommand.Type.STOP_NAVIGATION, null, cleaned)
        }

        // 4. 停止避障
        if (containsAny(cleaned, getKeywords("stop_obstacle"))) {
            return VoiceCommand(VoiceCommand.Type.STOP_OBSTACLE_AVOIDANCE, null, cleaned)
        }

        // 5. 开始避障
        if (containsAny(cleaned, getKeywords("start_obstacle"))) {
            return VoiceCommand(VoiceCommand.Type.START_OBSTACLE_AVOIDANCE, null, cleaned)
        }

        // 4. 我在哪里
        if (containsAny(cleaned, getKeywords("where_am_i"))) {
            return VoiceCommand(VoiceCommand.Type.WHERE_AM_I, null, cleaned)
        }

        // 5. 重复播报
        if (containsAny(cleaned, getKeywords("repeat"))) {
            return VoiceCommand(VoiceCommand.Type.REPEAT_LAST, null, cleaned)
        }

        // 6. 预览路线
        if (containsAny(cleaned, getKeywords("preview_route"))) {
            return VoiceCommand(VoiceCommand.Type.PREVIEW_ROUTE, null, cleaned)
        }

        // 7. 查询状态
        if (containsAny(cleaned, getKeywords("query_status"))) {
            return VoiceCommand(VoiceCommand.Type.QUERY_STATUS, null, cleaned)
        }

        // 8. 导航意图：提取目的地
        val destination = extractDestination(cleaned)
        if (!destination.isNullOrEmpty()) {
            Log.d(TAG, "Parsed NAVIGATE_TO destination: $destination")
            return VoiceCommand(VoiceCommand.Type.NAVIGATE_TO, destination, cleaned)
        }

        // 9. fallback：普通文本搜索
        if (cleaned.length >= 2) {
            Log.d(TAG, "Falling back to TEXT_SEARCH: $cleaned")
            return VoiceCommand(VoiceCommand.Type.TEXT_SEARCH, cleaned, cleaned)
        }

        Log.d(TAG, "Unknown command: $cleaned")
        return VoiceCommand(VoiceCommand.Type.UNKNOWN, null, cleaned)
    }

    private fun cleanText(text: String): String {
        return text.trim()
            .replace(Regex("^[。，、！；：,.!?;:]+"), "")
            .replace(Regex("[。，、！；：,.!?;:]+$"), "")
            .replace(Regex("^[吧吗呢啊]+"), "")
            .replace(Regex("[吧吗呢啊]+$"), "")
            .trim()
    }

    private fun containsAny(text: String, keywords: List<String>): Boolean {
        return keywords.any { text.contains(it) }
    }

    /**
     * 宽松匹配：精确匹配 + 单字子集兜底。
     *
     * 当百度 ASR 识别专有名词不可靠时（如"起义广场"→"起一广"），
     * 只要用户说的文本是某个关键词的连续子集，就算命中。
     * 例如 "起一广" ⊆ "起一广场" → true
     */
    private fun fuzzyMatch(text: String, keywords: List<String>): Boolean {
        if (containsAny(text, keywords)) return true

        // 子集兜底：用户说的每个字（忽略虚词）都出现在关键词中
        val meaningfulText = text.filter { it !in "去到我想的是了在吗呢吧啊" }.toSet()
        if (meaningfulText.size < 2) return false

        return keywords.any { kw ->
            val kwSet = kw.filter { it !in "去到我想的是了在吗呢吧啊" }.toSet()
            // 用户的字全部出现在关键词中 → 命中
            meaningfulText.all { it in kwSet }
        }
    }

    private fun extractDestination(text: String): String? {
        // 前缀匹配
        for (prefix in getKeywords("navigate_prefixes")) {
            if (text.startsWith(prefix)) {
                val dest = text.substring(prefix.length).trim()
                    .replace(Regex("^[的吧吗呢啊]+"), "")
                    .replace(Regex("[的吧吗呢啊]+$"), "")
                    .trim()
                if (dest.isNotEmpty()) return dest
            }
        }

        // 后缀匹配
        for (suffix in getKeywords("navigate_suffixes")) {
            if (text.endsWith(suffix)) {
                val dest = text.substring(0, text.length - suffix.length).trim()
                    .replace(Regex("^[的吧吗呢啊]+"), "")
                    .replace(Regex("[的吧吗呢啊]+$"), "")
                    .trim()
                if (dest.isNotEmpty()) return dest
            }
        }

        // 特殊处理："去XX"
        if (text.startsWith("去") && text.length > 1) {
            val dest = text.substring(1).trim()
                .replace(Regex("^[的吧吗呢啊]+"), "")
                .replace(Regex("[的吧吗呢啊]+$"), "")
                .trim()
            if (dest.isNotEmpty() && dest.length >= 2) return dest
        }

        // 特殊处理："到XX"
        if (text.startsWith("到") && text.length > 1) {
            val dest = text.substring(1).trim()
                .replace(Regex("^[的吧吗呢啊]+"), "")
                .replace(Regex("[的吧吗呢啊]+$"), "")
                .trim()
            if (dest.isNotEmpty() && dest.length >= 2) return dest
        }

        return null
    }
}
